deep_quoridor/src/quoridor.py

```diff
@@ -0,0 +1,293 @@
+        display_grid = np.full(self._grid.shape, " ", dtype=str)
+        display_grid[::2, ::2] = "."
+        display_grid[self._grid == Board.PLAYER1_WALL] = "w"
+        display_grid[self._grid == Board.PLAYER2_WALL] = "W"
+        for player, _ in enumerate(self._player_positions):
+            display_grid[*self._player_positions[player] * 2] = str(player + 1)
+        return "\n".join([" ".join(row) for row in display_grid]) + "\n"
+
+
+class Quoridor:
+    def __init__(self, board: Board):
+        """
+        If you want to start from the initial game state, pass in board=Board(board_size, max_walls).
+        """
+        self.board = board
+        self.current_player = 0
+
+        self._jump_checks = create_jump_checks()
+
+    def step(self, action: Action, validate: bool = True):
+        """
+        Execute the given action and update the game state.
+
+        Turning off validation can save some computation if you're sure the action is valid,
+        and also makes it easier to teleport players around the board for testing purposes.
+        """
+        if validate and not self.is_action_valid(action):
+            raise ValueError("Invalid action")
+
+        if isinstance(action, MoveAction):
+            self.board.move_player(self.current_player, action.destination)
+        elif isinstance(action, WallAction):
+            self.board.add_wall(self.current_player, action.position, action.orientation)
+            # TODO: Check that the wall doesn't block any player from reaching their goal.
+        else:
+            raise ValueError("Invalid action type")
+
+        self.current_player = 1 - self.current_player
+
+    def is_action_valid(self, action: Action):
+        """
+        Check whether the given action is valid given the current game state.
+        """
+        if isinstance(action, MoveAction):
+            player = self.get_current_player()
+            current_position = self.board.get_player_position(player)
+            opponent = 1 - player
+            opponent_position = self.board.get_player_position(opponent)
+            opponent_offset = opponent_position - current_position
+            destination_position = np.asarray(action.destination, dtype=np.uint8)
+            position_delta = tuple(destination_position - current_position)  # Tuple so we can use it as a key.
+
+            # Destination cell must be free
+            is_valid = True
+            if self.board.get_cell(destination_position) != Board.FREE:
+                is_valid = False
+
+            elif np.sum(np.abs(position_delta)) == 1:
+                # Moving to an adjacent cell, just make sure no wall is in the way.
+                if self.board.is_wall_between(current_position, destination_position):
+                    is_valid = False
+
+            elif (tuple(opponent_offset), tuple(position_delta)) in self._jump_checks:
+                jump_checks = self._jump_checks[(tuple(opponent_offset), tuple(position_delta))]
```

Comment:
You can use the walrus operator:
```suggestion
            elif (jump_checks := tuple(opponent_offset), tuple(position_delta)) in self._jump_checks:
```

Reply:
TIL!